lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -837,6 +837,65 @@ bool firrtl::areTypesWeaklyEquivalent(FIRRTLType destFType, FIRRTLType srcFType,
          destFlip == srcFlip;
 }
 
+/// Returns whether the srcType can be const-casted to the destType.
+bool firrtl::areTypesConstCastable(FIRRTLType destFType, FIRRTLType srcFType,
+                                   bool srcOuterTypeIsConst) {
+  // Identical types are always castable
+  if (destFType == srcFType)
+    return true;
+
+  auto destType = destFType.dyn_cast<FIRRTLBaseType>();
+  auto srcType = srcFType.dyn_cast<FIRRTLBaseType>();
+
+  // For non-base types, only castable if identical.
+  if (!destType || !srcType)
+    return false;
+
+  // Cannot cast non-'const' src to 'const' dest
+  if (destType.isConst() && !srcType.isConst() && !srcOuterTypeIsConst)
+    return false;
+
+  // Vector types can be casted if they have the same size and castable element
+  // type.
+  auto destVectorType = destType.dyn_cast<FVectorType>();
+  auto srcVectorType = srcType.dyn_cast<FVectorType>();
+  if (destVectorType && srcVectorType)
+    return destVectorType.getNumElements() == srcVectorType.getNumElements() &&
+           areTypesConstCastable(
+               destVectorType.getElementType(), srcVectorType.getElementType(),
+               srcOuterTypeIsConst || srcVectorType.isConst());
+
+  // Bundle types can be casted if they have the same size, element names,
+  // and castable element types.
+  auto destBundleType = destType.dyn_cast<BundleType>();
+  auto srcBundleType = srcType.dyn_cast<BundleType>();
+  if (destBundleType && srcBundleType) {
+    auto destElements = destBundleType.getElements();
+    auto srcElements = srcBundleType.getElements();
+    size_t numDestElements = destElements.size();
+    if (numDestElements != srcElements.size())
+      return false;
+
+    for (size_t i = 0; i < numDestElements; ++i) {
+      auto destElement = destElements[i];
+      auto srcElement = srcElements[i];
+      if (destElement.name != srcElement.name ||
+          destElement.isFlip != srcElement.isFlip ||
+          !areTypesConstCastable(destElement.type, srcElement.type,
+                                 srcOuterTypeIsConst ||
+                                     srcBundleType.isConst()))
+        return false;
+    }
+    return true;
+  }
+
+  // Ground types can be casted if the source type is a const
+  // version of the destination type
+  return destType == srcType.getConstType(false) ||
+         (destType.isConst() && srcOuterTypeIsConst &&
```

Comment:
Following the "srcIsEffectivelyConst" idea, this seems like it maybe "should" be `(srcOuterTypeIsConst || srcType.isConst())` but doesn't need to check for `srcType.isConst()` because if srcType was const then we'd have exited earlier with types already being equivalent?

Reply:
I've simplified the final check